File: Source/Sql/Engine/Engine.cpp
#include "../Engine/Engine.h"

#include <utility>

#include <absl/status/status.h>

#include "../Parser/Parser.h"
#include "../Plan/Planner.h"

namespace spacedb
{
    Session::Session(ISqlEngine& engine) : engine_(engine)
    {
    }

    absl::StatusOr<ResultSet> Session::Execute(std::string_view sql)
    {
        // 第一步：词法/语法解析 → AST
        absl::StatusOr<Statement> statement = Parser(sql).Parse();
        if (!statement.ok())
        {
            return statement.status();
        }

        // 第二步：规划 → 计划树
        absl::StatusOr<Plan> plan = Planner::Build(std::move(*statement));
        if (!plan.ok())
        {
            return plan.status();
        }

        // 第三步：开启事务
        absl::StatusOr<std::unique_ptr<ISqlTransaction>> txn = engine_.Begin();
        if (!txn.ok())
        {
            return txn.status();
        }

        // 第四步：在事务中执行计划
        absl::StatusOr<ResultSet> result = plan->Execute(**txn);
        if (!result.ok())
        {
            // 第五步：失败 → 回滚，向上传播原始错误。
            // 回滚自身失败的细节暂不叠加（阶段 8 MVCC 再细化）。
            std::ignore = (*txn)->Rollback();
            return result.status();
        }

        // 第六步：成功 → 提交
        if (absl::Status status = (*txn)->Commit(); !status.ok())
        {
            return status;
        }
        return result;
    }
} // namespace spacedb

File: engine/engine.go
package engine

import (
	"errors"
	"fmt"
	"spacedb/executor"
	"spacedb/parser"
	"spacedb/planner"
)

// Engine 是 SQL 引擎依赖的接口
//
// 具体实现可以是内存引擎、KV 引擎或分布式引擎,
// Engine 不关心 SQL 解析和执行细节，只负责开启事务
type Engine interface {
	Begin() (executor.Transaction, error)
}

// Session 表示一个客户端会话
//
// Session 持有 Engine，但不直接持有事务,
// 每次 Execute 都创建一个新的事务
type Session struct {
	engine Engine
}

func NewSession(engine Engine) *Session {
	return &Session{engine: engine}
}

// Execute 执行一条 SQL 语句
func (s *Session) Execute(sql string) (executor.ResultSet, error) {
	// 解析 AST
	stmt, err := parser.Parse(sql)
	if err != nil {
		return nil, fmt.Errorf("engine: parsing SQL: %w", err)
	}

	// 构建执行计划节点
	plan, err := planner.Build(stmt)
	if err != nil {
		return nil, fmt.Errorf("engine: building plan: %w", err)
	}

	// 每个节点都有一个执行器对应
	exec, err := executor.Build(plan.Node)
	if err != nil {
		return nil, fmt.Errorf("engine: building executor: %w", err)
	}

	// 从 存储引擎获取 txn
	txn, err := s.engine.Begin()
	if err != nil {
		return nil, fmt.Errorf("engine: beginning transaction: %w", err)
	}

	// 将 txn 交给执行器执行
	result, err := exec.Execute(txn)
	if err != nil {
		if rollbackErr := txn.Rollback(); rollbackErr != nil {
			return nil, errors.Join(
				fmt.Errorf("engine: executing statement: %w", err),
				fmt.Errorf("engine: rolling back transaction: %w", rollbackErr),
			)
		}
		return nil, fmt.Errorf("engine: executing statement: %w", err)
	}

	if err := txn.Commit(); err != nil {
		return nil, fmt.Errorf("engine: committing transaction: %w", err)
	}

	return result, nil
}
